### PR TITLE
TRUNK-5835:OrderSetAttribute implementation

### PR DESCRIPTION
**Pull request by gitcliff**
link:https://issues.openmrs.org/browse/TRUNK-5835

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/OrderSetService.java [hunk 1]
```diff
@@ -107,4 +109,99 @@
...
+	 */
+	@Authorized(PrivilegeConstants.MANAGE_ORDER_SET_ATTRIBUTE_TYPES)
+	OrderSetAttributeType saveOrderSetAttributeType(OrderSetAttributeType orderSetAttributeType);
+
+	/**
+	 * Retires the given orderSet attribute type in the database
+	 * 
+	 * @param orderSetAttributeType
+	 * @return the orderSetAttribute retired
+	 * @since 1.9
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/OrderSetService.java:167]: Since what?

Comment (gitcliff) [reply api/src/main/java/org/openmrs/api/OrderSetService.java:167]: oh sorry,,,my bad ,gonna fix it

Comment (gitcliff) [reply api/src/main/java/org/openmrs/api/OrderSetService.java]: i have fixed it

**Comment by dkayiwa** [comment]
Can you update DatabaseUpdaterDatabaseIT with the changesets in this pull request?

**Comment by dkayiwa** [comment]
You also do not seem to have implemented all the changes in the green commits here: https://github.com/idlewis/openmrs-core/commits/order_set_attributes

**Comment by gitcliff** [comment]
@dkayiwa i have added some changes ,,please review

**Comment by dkayiwa** [comment]
The ticket status does not indicate that you are ready for review.

**Comment by dkayiwa** [comment]
@gitcliff did you do all the changes as required by the ticket description?

**Comment by gitcliff** [comment]
@dkayiwa i have added some changes

**Comment by dkayiwa** [comment]
Did you see the travis build failure?

**Comment by gitcliff** [comment]
@dkayiwa i looked into the travis failure and still trying to figure out what is the cause because all the tests are passing locally on my pc ......
Though when i look into travis i see the problem around `DatabaseUpdaterDatabaseIT` changeset numbering

**Comment by dkayiwa** [comment]
The error message is clear enough for you to know what to do.

**Comment by gitcliff** [comment]
@dkayiwa i decided to first maintain the DatabaseUpdaterDatabaseIT changesets number to the one currently in core in order to first deal with why the` shouldValidateHibernateMappings` test is failing in `ValidateHibernateMappingsDatabaseIT` test class with `org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing table [order_set_attribute]` yet the `order_set_attribute `table has been created as seen in the attached screenshots

**Comment by dkayiwa** [comment]
@gitcliff to reproduce this locally, just run `mvn clean test -Pskip-default-test -Pintegration-test`

**Comment by dkayiwa** [comment]
@gitcliff the error you are getting is cause by your restricting the changesets to MySQL with `dbms="mysql"`. So remove it.

**Comment by gitcliff** [comment]
@dkayiwa  i have removed` dbms="mysql"` from the changesets though still facing some errors here https://pastebin.com/jbhnJ6Y4

**Comment by dkayiwa** [comment]
Can you fix them too?

**Comment by gitcliff** [comment]
@dkayiwa i have fixed them,,requesting for review

**Comment by dkayiwa** [comment]
You still have not implemented all the changes in the green commits here: https://github.com/idlewis/openmrs-core/commits/order_set_attributes

**Comment by gitcliff** [comment]
@dkayiwa requesting for review

File: api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.4.x.xml [hunk 2]
```diff
@@ -184,4 +185,79 @@
...
+				<constraints nullable="false" />
+			</column>
+			<column name="changed_by" type="int" />
+			<column name="date_changed" type="datetime" />
+			<column name="retired" type="boolean" defaultValueBoolean="false">
+				<constraints nullable="false" />
+			</column>
+			<column name="retired_by" type="int" />
+			<column name="date_retired" type="datetime" />
+			<column name="retire_reason" type="varchar(255)" defaultValue="null" />
```
Comment (dkayiwa) [inline api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.4.x.xml:223]: Do you really need a default value of null?

File: api/src/main/java/org/openmrs/api/OrderSetService.java [hunk 3]
```diff
@@ -107,4 +109,99 @@
 	@Authorized(PrivilegeConstants.GET_ORDER_SETS)
 	OrderSetMember getOrderSetMemberByUuid(String uuid);
 
+	/**
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/OrderSetService.java:112]: Add a method description.

File: api/src/main/java/org/openmrs/api/OrderSetService.java [hunk 4]
```diff
@@ -107,4 +109,99 @@
...
+	 * @param uuid
+	 * @return the {@link OrderSetAttribute} with the given uuid
+	 * @since 2.4.0
+	 * @should get the order set attribute with the given uuid
+	 * @should return null if no order set attribute has the given uuid
+	 */
+	@Authorized(PrivilegeConstants.GET_ORDER_SETS)
+	OrderSetAttribute getOrderSetAttributeByUuid(String uuid);
+
+	/**
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/OrderSetService.java:122]: Add a method description.

File: api/src/main/java/org/openmrs/api/OrderSetService.java [hunk 5]
```diff
@@ -107,4 +109,99 @@
...
+	 * @return all {@link OrderSetAttributeType}s
+	 * @since 2.4.0
+	 * @should return all orderSet attribute types including retired ones
+	 */
+	@Authorized(PrivilegeConstants.GET_ORDER_SET_ATTRIBUTE_TYPES)
+	List<OrderSetAttributeType> getAllOrderSetAttributeTypes();
+
+	/**
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/OrderSetService.java:130]: Add a method description.

File: api/src/main/java/org/openmrs/api/OrderSetService.java [hunk 6]
```diff
@@ -107,4 +109,99 @@
...
+	 * @param id
+	 * @return the {@link OrderSetAttributeType} with the given internal id
+	 * @since 2.4.0
+	 * @should return the orderSet attribute type with the given id
+	 * @should return null if no orderSet attribute type exists with the given id
+	 */
+	@Authorized(PrivilegeConstants.GET_ORDER_SET_ATTRIBUTE_TYPES)
+	OrderSetAttributeType getOrderSetAttributeType(Integer id);
+
+	/**
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/OrderSetService.java:140]: Add a method description.

File: api/src/main/java/org/openmrs/api/OrderSetService.java [hunk 7]
```diff
@@ -107,4 +109,99 @@
 	@Authorized(PrivilegeConstants.GET_ORDER_SETS)
 	OrderSetMember getOrderSetMemberByUuid(String uuid);
 
+	/**
+	 * @param uuid
+	 * @return the {@link OrderSetAttribute} with the given uuid
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/OrderSetService.java:114]: Can you add a test method for each @should annotation?

File: api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.4.x.xml [hunk 8]
```diff
@@ -184,4 +185,79 @@
...
+            <column name="creator" type="int"><constraints nullable="false"/></column>
+            <column name="date_created" type="datetime"><constraints nullable="false"/></column>
+            <column name="changed_by" type="int"/>
+            <column name="date_changed" type="datetime" />
+            <column name="voided" type="boolean" defaultValueBoolean="false">
+                <constraints nullable="false"/>
+            </column>
+            <column name="voided_by" type="int" />
+            <column name="date_voided" type="datetime" />
+            <column name="void_reason" type="varchar(255)" defaultValue="null"/>
```
Comment (dkayiwa) [inline api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.4.x.xml:253]: Do you really need a default value of null?

File: api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.4.x.xml (hunk 2 above)
Comment (gitcliff) [reply api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.4.x.xml:223]: @dkayiwa i kept it to maintain the way its done in other domains that implement attribute

Comment (dkayiwa) [reply api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.4.x.xml:223]: Like which other domains?

**Pull request merged**